ports/stm32f4/common-hal/busio/I2C.c

```diff
@@ -32,52 +32,102 @@
 
 #include "shared-bindings/microcontroller/__init__.h"
 #include "supervisor/shared/translate.h"
-
-I2C_HandleTypeDef hi2c2;
+#include "common-hal/microcontroller/Pin.h"
+#include "stm32f4/periph.h"
+
+STATIC I2C_HandleTypeDef hi2c;
+
+STATIC I2C_TypeDef * I2Cbanks[3] = 
+{
+#ifdef I2C1
+    I2C1, 
+#else 
+    NULL,
+#endif
+#ifdef I2C2
+    I2C2,
+#else
+    NULL,
+#endif
+#ifdef I2C3
+    I2C3
+#else
+    NULL
+#endif
+};
 
 void common_hal_busio_i2c_construct(busio_i2c_obj_t *self,
         const mcu_pin_obj_t* scl, const mcu_pin_obj_t* sda, uint32_t frequency, uint32_t timeout) {
-    //TODO: Rework this entire section to use LL so we can properly assign pins
-    //      This will also be bundled with MSP removal
-    hi2c2.Instance = I2C1;
-    hi2c2.Init.ClockSpeed = 100000;
-    hi2c2.Init.DutyCycle = I2C_DUTYCYCLE_2;
-    hi2c2.Init.OwnAddress1 = 0;
-    hi2c2.Init.AddressingMode = I2C_ADDRESSINGMODE_7BIT;
-    hi2c2.Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
-    hi2c2.Init.OwnAddress2 = 0;
-    hi2c2.Init.GeneralCallMode = I2C_GENERALCALL_DISABLE;
-    hi2c2.Init.NoStretchMode = I2C_NOSTRETCH_DISABLE;
-    if(HAL_I2C_Init(&hi2c2) != HAL_OK) {
+
+    I2C_TypeDef * I2Cx;
+    uint8_t i2cidx = 0;
+    uint8_t i2cidx_c = 0;
+    uint8_t len = sizeof(mcu_i2c_list)/sizeof(*mcu_i2c_list);
+    for(uint i=0; i<len;i++) {
+        if (mcu_i2c_list[i]->sda_pin == sda) {
+            i2cidx = mcu_i2c_list[i]->i2c_index;
+        }
+        if (mcu_i2c_list[i]->scl_pin == scl) {
+            i2cidx_c = mcu_i2c_list[i]->i2c_index;
+        }
+    }
+    if(i2cidx == i2cidx_c) {
+        if(I2Cbanks[i2cidx] == NULL || i2cidx>=len) {
+            mp_raise_RuntimeError(translate("Bad I2C pin definition, check peripheral files"));
+        }
+        I2Cx = I2Cbanks[i2cidx];
+    } else {
+        mp_raise_RuntimeError(translate("Invalid I2C pin selection"));
+    } 
+
+    GPIO_InitTypeDef GPIO_InitStruct = {0};
+    GPIO_InitStruct.Pin = pin_mask(sda->number)|pin_mask(scl->number);
+    GPIO_InitStruct.Mode = GPIO_MODE_AF_OD;
+    GPIO_InitStruct.Pull = GPIO_PULLUP;
+    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
+    GPIO_InitStruct.Alternate = GPIO_AF4_I2C1; //TODO; custom per pin? Or error if mixed is bad.
```

Comment:
Mixed should be fine. In SAMD since we track options per pin, we have three sercom slots ordered by their AF number.

Reply:
Sorry, I guess I really shouldn't push unless I'm finished with an approach, so you aren't needlessly reviewing WIP code. My new system reworks most of this.